# Importar librerias y data sets

In [1]:
# importar librerías
import pandas as pd
import numpy as np

In [2]:
# cargar archivos
orders = pd.read_csv("https://practicum-content.s3.amazonaws.com/datasets/rappiplus_orders_raw.csv")
catalog = pd.read_csv("https://practicum-content.s3.amazonaws.com/datasets/rappiplus_catalog.csv")
marketing = pd.read_csv("https://practicum-content.s3.amazonaws.com/datasets/rappiplus_marketing_spend.csv")

In [3]:
# explorar datasets
print("Orders:")
orders.info()
print() #salto de linea
print("Catalog:")
catalog.info()
print() #salto de linea
print("Marketing")
marketing.info()

Orders:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25100 entries, 0 to 25099
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id_pedido           25100 non-null  object 
 1   id_usuario          25100 non-null  object 
 2   fecha_hora_pedido   25100 non-null  object 
 3   pais                24800 non-null  object 
 4   dispositivo         25080 non-null  object 
 5   fuente_referencia   25070 non-null  object 
 6   nombre_producto     25070 non-null  object 
 7   categoria_producto  25020 non-null  object 
 8   cantidad            25050 non-null  float64
 9   precio_unitario     25050 non-null  float64
 10  monto_descuento     25050 non-null  float64
 11  monto_total         25100 non-null  float64
dtypes: float64(4), object(8)
memory usage: 2.3+ MB

Catalog:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column              Non-Null C

In [4]:
# explorar datasets
print("Orders:")
orders.head(-10)

Orders:


,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2.0,332.69,0.0,665.37
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,176.86,5.0,171.86
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2.0,102.99,10.0,195.99
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electronica,1.0,257.87,15.0,242.87
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1.0,336.28,0.0,336.28
...,...,...,...,...,...,...,...,...,...,...,...,...
25085,order_21655,user_4706,2025-05-18,Mexico,desktop,social,Laptop-Gaming-16GB,Electronica,2.0,57.46,5.0,109.92
25086,order_21527,user_3071,2025-05-30,Argentina,desktop,organic,Blender-XL-Red,Hogar,2.0,294.72,0.0,589.44
25087,order_18327,user_5721,2025-04-23,Argentina,mobile,paid_search,Vacuum-Pro-Black,Hogar,1.0,38.58,0.0,38.58
25088,order_20453,user_1237,2025-04-16,Colombia,desktop,organic,Phone-Pro-128GB,Electronica,1.0,461.93,0.0,461.93


In [5]:
print("Catalog:")
catalog.head(7)

Catalog:


,nombre_producto,categoria_producto,costo_unitario,proveedor
0,Laptop-Gaming-16GB,Electrónica,280.68,"Fuller, Pena and Myers"
1,Phone-Pro-128GB,Electrónica,10.12,King Ltd
2,Tablet-Standard-64GB,Electrónica,25.21,Bowers LLC
3,Blender-XL-Red,Hogar,176.64,Long-Reid
4,Vacuum-Pro-Black,Hogar,16.60,"Rivera, Carr and Finley"
5,Sneakers-Urban-42,Moda,17.21,Greene-Smith
6,Jacket-Winter-M,Moda,189.31,Mcmillan-Rhodes


In [6]:
print("Marketing:")
marketing.head(10)

Marketing:


,fecha,pais,id_campaña,canal,gasto
0,2025-01-01,Mexico,organic_Mexico,organic,2446.25
1,2025-01-01,Mexico,paid_search_Mexico,paid_search,2704.34
2,2025-01-01,Mexico,social_Mexico,social,2045.01
3,2025-01-01,Colombia,organic_Colombia,organic,2597.21
4,2025-01-01,Colombia,paid_search_Colombia,paid_search,1771.40
5,2025-01-01,Colombia,social_Colombia,social,833.38
6,2025-01-01,Argentina,organic_Argentina,organic,1365.62
7,2025-01-01,Argentina,paid_search_Argentina,paid_search,1282.65
8,2025-01-01,Argentina,social_Argentina,social,1534.69
9,2025-01-02,Mexico,organic_Mexico,organic,2537.38


# Data Frame: Catalog

In [7]:
# Unificación de nombres en categoria_producto: Electrónica -----> Electronica
catalog['categoria_producto'] = catalog['categoria_producto'].replace({'Electrónica': 'Electronica'})

catalog.head(7)

,nombre_producto,categoria_producto,costo_unitario,proveedor
0,Laptop-Gaming-16GB,Electronica,280.68,"Fuller, Pena and Myers"
1,Phone-Pro-128GB,Electronica,10.12,King Ltd
2,Tablet-Standard-64GB,Electronica,25.21,Bowers LLC
3,Blender-XL-Red,Hogar,176.64,Long-Reid
4,Vacuum-Pro-Black,Hogar,16.60,"Rivera, Carr and Finley"
5,Sneakers-Urban-42,Moda,17.21,Greene-Smith
6,Jacket-Winter-M,Moda,189.31,Mcmillan-Rhodes


# Data Frame: Orders

In [8]:
# Cambio de formato object------>pd.to_datetime() para la columna de fechas para ambos df: orders y marketing
orders['fecha_hora_pedido'] = pd.to_datetime(orders['fecha_hora_pedido'])
marketing['fecha'] = pd.to_datetime(marketing['fecha'])

In [9]:
print("Estadistica Tabla Orders:")
orders.describe()

Estadistica Tabla Orders:


,cantidad,precio_unitario,monto_descuento,monto_total
count,25050.000000,25050.000000,25050.000000,2.510000e+04
mean,7.092735,259.305549,4.500798,2.072680e+03
std,296.277003,138.726461,5.223010,9.894995e+04
min,-2.000000,20.030000,0.000000,-4.926500e+02
25%,1.000000,138.377500,0.000000,1.805075e+02
50%,2.000000,258.715000,0.000000,3.417500e+02
75%,2.000000,380.332500,10.000000,5.185800e+02
max,20000.000000,499.960000,15.000000,8.840200e+06


In [10]:
# Valores únicos de la columna id_pedido
print("Total de transacciones únicas:", orders['id_pedido'].nunique() )
print("Total de filas repetidas:", orders.duplicated().sum())

Total de transacciones únicas: 25000
Total de filas repetidas: 100


In [11]:
# Reemplazar filas repetidas 
orders = orders.drop_duplicates()

# Comrpobación del cambio
print(orders.shape[0])

25000


In [12]:
len(orders[orders['cantidad'].abs() >= 100])

10

In [13]:
# Verificación de cantidad de piezas con potenciales errores de captura (ejem. 10,000 a 20,000 pzs)
orders[orders['cantidad']>= 100]

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
3521,order_3521,user_5812,2025-02-03,Mexico,mobile,paid_search,Laptop-Gaming-16GB,Electronica,10000.0,43.14,0.0,431400.0
3522,order_3522,user_3575,2025-03-29,Argentina,desktop,social,Laptop-Gaming-16GB,Electronica,10000.0,280.55,0.0,2805500.0
3586,order_3586,user_3380,2025-02-03,Mexico,mobile,paid_search,Laptop-Gaming-16GB,Electronica,10000.0,490.35,0.0,4903500.0
3643,order_3643,user_4440,2025-01-07,Colombia,desktop,social,Laptop-Gaming-16GB,Electronica,10000.0,238.15,0.0,2381500.0
3656,order_3656,user_884,2025-01-01,Argentina,mobile,organic,Laptop-Gaming-16GB,Electronica,20000.0,297.66,0.0,5953200.0
3668,order_3668,user_7270,2025-06-24,Mexico,mobile,paid_search,Laptop-Gaming-16GB,Electronica,20000.0,348.31,0.0,6966200.0
3689,order_3689,user_6566,2025-06-16,mexico,desktop,paid_search,Laptop-Gaming-16GB,Electronica,10000.0,87.69,0.0,876900.0
3722,order_3722,user_4723,2025-05-09,Argentina,mobile,paid_search,Laptop-Gaming-16GB,Electronica,20000.0,442.01,0.0,8840200.0
3726,order_3726,user_2536,2025-02-12,Colombia,desktop,social,Laptop-Gaming-16GB,Electronica,20000.0,290.85,0.0,5817000.0
3748,order_3748,user_7096,2025-02-23,Mexico,desktop,organic,Laptop-Gaming-16GB,Electronica,10000.0,336.93,0.0,3369300.0


In [14]:
orders.shape

(25000, 12)

In [15]:

# Comprobación de filas con numeros en la columna cantidad fuera de lo común
filas_cantidad_superior= len(orders[orders['cantidad']>= 100])
print(f' Transacciones con cantidad de piezas mayor o igual a cien: {filas_cantidad_superior}') 


 Transacciones con cantidad de piezas mayor o igual a cien: 10


In [16]:
# Comprobación de filas con numeros en la columna cantidad fuera de lo común
nulos_cantidad= len(orders[orders['cantidad'].isna() ])
print(f' Total de valores nulos en la columna canitdad: {nulos_cantidad}') 


 Total de valores nulos en la columna canitdad: 50


In [17]:
# Nulos en cantidad
orders[orders['cantidad'].isna()]

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
74,order_74,user_6172,2025-01-15,Argentina,desktop,organic,Sneakers-Urban-42,NaN,NaN,NaN,NaN,595.85
75,order_75,user_6588,2025-06-19,Colombia,mobile,organic,Sneakers-Urban-42,NaN,NaN,NaN,NaN,458.15
76,order_76,user_3193,2025-03-17,Argentina,mobile,paid_search,Laptop-Gaming-16GB,NaN,NaN,NaN,NaN,319.75
77,order_77,user_775,2025-06-24,Argentina,desktop,social,Vacuum-Pro-Black,NaN,NaN,NaN,NaN,227.55
78,order_78,user_2702,2025-03-19,Argentina,desktop,paid_search,Phone-Pro-128GB,NaN,NaN,NaN,NaN,432.39
79,order_79,user_6438,2025-03-22,Colombia,desktop,paid_search,Sneakers-Urban-42,NaN,NaN,NaN,NaN,527.15
80,order_80,user_6790,2025-01-03,Mexico,desktop,organic,Vacuum-Pro-Black,NaN,NaN,NaN,NaN,711.18
81,order_81,user_232,2025-05-16,Mexico,mobile,paid_search,Tablet-Standard-64GB,NaN,NaN,NaN,NaN,41.08
82,order_82,user_5665,2025-02-07,Colombia,mobile,paid_search,Sneakers-Urban-42,NaN,NaN,NaN,NaN,431.55
83,order_83,user_5225,2025-04-25,Argentina,mobile,organic,Sneakers-Urban-42,NaN,NaN,NaN,NaN,279.35


In [18]:

# Conservar valores absolutos < 100 Y TAMBIÉN los nulos (NaN) para contemplar la información de las columnas categóricas
orders = orders[( orders['cantidad'].abs() < 100 ) | ( orders['cantidad'].isna() )].copy()

In [19]:
# Validación de valores menor o igual a cero (Se consideran devoluciones, por lo que se mantienen intactos)
filas_con_negativos = orders[(orders['cantidad'] <= 0) | (orders['monto_total'] <= 0)]
filas_con_negativos

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
266,order_266,user_7011,2025-03-13,NaN,desktop,paid_search,Phone-Pro-128GB,Electronica,-2.0,101.31,10.0,-192.62
267,order_267,user_1087,2025-05-07,NaN,desktop,social,Phone-Pro-128GB,Electronica,-1.0,43.50,5.0,-38.50
268,order_268,user_84,2025-02-19,NaN,desktop,organic,Phone-Pro-128GB,Electronica,-1.0,497.65,5.0,-492.65
269,order_269,user_3323,2025-05-25,NaN,desktop,paid_search,Phone-Pro-128GB,Electronica,-1.0,423.53,0.0,-423.53


In [20]:
# Nueva columna para determinar que tipo de transacción segun la columna monto_total (Venta [> 0] ó Devolución [<0]) para facilitar el análisis en Power BI
orders['tipo_transaccion'] = orders['monto_total'].apply(lambda x: 'Devolucion' if x < 0 else 'Venta')

orders.head()

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total,tipo_transaccion
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2.0,332.69,0.0,665.37,Venta
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,176.86,5.0,171.86,Venta
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2.0,102.99,10.0,195.99,Venta
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electronica,1.0,257.87,15.0,242.87,Venta
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1.0,336.28,0.0,336.28,Venta


In [21]:
# Total de valores nulos por columna
orders.isna().sum()

id_pedido               0
id_usuario              0
fecha_hora_pedido       0
pais                  300
dispositivo            20
fuente_referencia      30
nombre_producto        30
categoria_producto     80
cantidad               50
precio_unitario        50
monto_descuento        50
monto_total             0
tipo_transaccion        0
dtype: int64

In [22]:
# Porcentaje de valores nulos por columna
valores_nulos= orders.isna().sum()
pct_nulos = (valores_nulos / orders.shape[0]) * 100

print("Proporción de valores nulos por columna (%):")
pct_nulos

Proporción de valores nulos por columna (%):


id_pedido             0.000000
id_usuario            0.000000
fecha_hora_pedido     0.000000
pais                  1.200480
dispositivo           0.080032
fuente_referencia     0.120048
nombre_producto       0.120048
categoria_producto    0.320128
cantidad              0.200080
precio_unitario       0.200080
monto_descuento       0.200080
monto_total           0.000000
tipo_transaccion      0.000000
dtype: float64

In [23]:
orders[orders['categoria_producto'].isna()]

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total,tipo_transaccion
44,order_44,user_2899,2025-02-22,Colombia,desktop,NaN,NaN,NaN,1.0,318.38,5.0,313.38,Venta
45,order_45,user_6196,2025-02-10,Mexico,mobile,NaN,NaN,NaN,2.0,354.06,10.0,698.12,Venta
46,order_46,user_5815,2025-04-24,Mexico,mobile,NaN,NaN,NaN,1.0,359.31,10.0,349.31,Venta
47,order_47,user_406,2025-05-08,Colombia,mobile,NaN,NaN,NaN,2.0,476.78,0.0,953.56,Venta
48,order_48,user_6187,2025-01-26,Colombia,desktop,NaN,NaN,NaN,2.0,137.96,0.0,275.93,Venta
...,...,...,...,...,...,...,...,...,...,...,...,...,...
119,order_119,user_4494,2025-01-16,Argentina,mobile,social,Sneakers-Urban-42,NaN,NaN,NaN,NaN,196.45,Venta
120,order_120,user_4220,2025-02-03,Mexico,mobile,organic,Sneakers-Urban-42,NaN,NaN,NaN,NaN,355.15,Venta
121,order_121,user_2114,2025-04-25,colombia,desktop,social,Sneakers-Urban-42,NaN,NaN,NaN,NaN,724.38,Venta
122,order_122,user_2546,2025-06-23,Mexico,desktop,paid_search,Sneakers-Urban-42,NaN,NaN,NaN,NaN,469.20,Venta


In [24]:
# Reemplazar valores NaN en las columnas categóricas por "Sin especificar"
cols_categoricas = ['pais', 'dispositivo', 'fuente_referencia', 'nombre_producto', 'categoria_producto']

orders[cols_categoricas] = orders[cols_categoricas].fillna("Sin especificar")

In [25]:
# Unificar formato de nombres de los paises 
orders['pais']= orders['pais'].replace( {'mexico': 'Mexico', 'argentina': 'Argentina', 'colombia': 'Colombia'} )

print("Comprobqción de cambios:", orders['pais'].unique() )

Comprobqción de cambios: ['Argentina' 'Mexico' 'Colombia' 'Sin especificar']


In [26]:
# Verificación de valores unicos en el resto de columnas categoricas
print( "Nombres únicos de dispositivo:", orders['dispositivo'].unique() )
print() #salto de linea
print( "Nombres únicos de fuente_referencia:", orders['fuente_referencia'].unique() )
print() #Salto de linea
print( "Nombres únicos de nombre_producto:", orders['nombre_producto'].unique() )
print() #Salto de linea
print( "Nombres únicos de categoria_producto:", orders['categoria_producto'].unique() )

Nombres únicos de dispositivo: ['desktop' 'mobile' 'Sin especificar']

Nombres únicos de fuente_referencia: ['organic' 'paid_search' 'social' 'Sin especificar']

Nombres únicos de nombre_producto: ['Jacket-Winter-M' 'Tablet-Standard-64GB' 'Blender-XL-Red'
 'Laptop-Gaming-16GB' 'Sneakers-Urban-42' 'Phone-Pro-128GB'
 'Vacuum-Pro-Black' 'Sin especificar']

Nombres únicos de categoria_producto: ['Moda' 'Electronica' 'Hogar' 'Sin especificar']


In [27]:
# Comprobación de cambios en columnas categóricas
orders.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24990 entries, 0 to 24999
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   id_pedido           24990 non-null  object        
 1   id_usuario          24990 non-null  object        
 2   fecha_hora_pedido   24990 non-null  datetime64[ns]
 3   pais                24990 non-null  object        
 4   dispositivo         24990 non-null  object        
 5   fuente_referencia   24990 non-null  object        
 6   nombre_producto     24990 non-null  object        
 7   categoria_producto  24990 non-null  object        
 8   cantidad            24940 non-null  float64       
 9   precio_unitario     24940 non-null  float64       
 10  monto_descuento     24940 non-null  float64       
 11  monto_total         24990 non-null  float64       
 12  tipo_transaccion    24990 non-null  object        
dtypes: datetime64[ns](1), float64(4), object(8)
me

In [28]:
# Crear mapeo {nombre_producto: categoria_producto} a partir de catalog
mapa_categorias = catalog.set_index('nombre_producto')['categoria_producto']

# Actualización directa sobre la columna en orders
orders['categoria_producto'] = orders['nombre_producto'].map(mapa_categorias)

# Comprobación de cambios
orders[orders['cantidad'].isna()]

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total,tipo_transaccion
74,order_74,user_6172,2025-01-15,Argentina,desktop,organic,Sneakers-Urban-42,Moda,NaN,NaN,NaN,595.85,Venta
75,order_75,user_6588,2025-06-19,Colombia,mobile,organic,Sneakers-Urban-42,Moda,NaN,NaN,NaN,458.15,Venta
76,order_76,user_3193,2025-03-17,Argentina,mobile,paid_search,Laptop-Gaming-16GB,Electronica,NaN,NaN,NaN,319.75,Venta
77,order_77,user_775,2025-06-24,Argentina,desktop,social,Vacuum-Pro-Black,Hogar,NaN,NaN,NaN,227.55,Venta
78,order_78,user_2702,2025-03-19,Argentina,desktop,paid_search,Phone-Pro-128GB,Electronica,NaN,NaN,NaN,432.39,Venta
79,order_79,user_6438,2025-03-22,Colombia,desktop,paid_search,Sneakers-Urban-42,Moda,NaN,NaN,NaN,527.15,Venta
80,order_80,user_6790,2025-01-03,Mexico,desktop,organic,Vacuum-Pro-Black,Hogar,NaN,NaN,NaN,711.18,Venta
81,order_81,user_232,2025-05-16,Mexico,mobile,paid_search,Tablet-Standard-64GB,Electronica,NaN,NaN,NaN,41.08,Venta
82,order_82,user_5665,2025-02-07,Colombia,mobile,paid_search,Sneakers-Urban-42,Moda,NaN,NaN,NaN,431.55,Venta
83,order_83,user_5225,2025-04-25,Argentina,mobile,organic,Sneakers-Urban-42,Moda,NaN,NaN,NaN,279.35,Venta


# Data Frame: Marketing

In [29]:
# Corrección de campos de id_cammpaña (p. ejm: organic_Mexico----> organic) para evitar errores de filtrado en software de BI
marketing['id_campaña'] = marketing['id_campaña'].str.rsplit('_', n=1).str[0]

In [30]:
# Campos únicos en la columna id_campaña
marketing['id_campaña'].unique()

array(['organic', 'paid_search', 'social'], dtype=object)

In [31]:
marketing[marketing['canal'].isna()]

,fecha,pais,id_campaña,canal,gasto
98,2025-01-11,Argentina,social,NaN,849.70
99,2025-01-12,Mexico,organic,NaN,2033.56
100,2025-01-12,Mexico,paid_search,NaN,1260.65
101,2025-01-12,Mexico,social,NaN,1660.90
102,2025-01-12,Colombia,organic,NaN,1819.27
...,...,...,...,...,...
194,2025-01-22,Colombia,social,NaN,772.46
195,2025-01-22,Argentina,organic,NaN,1562.61
196,2025-01-22,Argentina,paid_search,NaN,644.34
197,2025-01-22,Argentina,social,NaN,2544.66


In [32]:
# Reemplazar valores nulos con "Sin especificar" en la columna canal 
marketing['canal']= marketing['canal'].fillna("Sin especificar")

# Comprobación de valores nuklos despues de la imputación.
print( "Valores nulos despues de imputación:", marketing['canal'].isna().sum() )

Valores nulos despues de imputación: 0


In [33]:
print("Estadistica Tabla Marketing:")
marketing.describe()

Estadistica Tabla Marketing:


,gasto
count,1620.00000
mean,1772.74292
std,734.43294
min,501.11000
25%,1128.03000
50%,1782.42500
75%,2420.68500
max,2999.36000


**Comprobación final**

In [34]:
# Comprobción final de los df orders y marketing
print("Orders:")
orders.info()
print() #salto de linea
print("Marketing:")
marketing.info()
print() #Salto de linea
print("Catalog:")
catalog.info()

Orders:
<class 'pandas.core.frame.DataFrame'>
Int64Index: 24990 entries, 0 to 24999
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   id_pedido           24990 non-null  object        
 1   id_usuario          24990 non-null  object        
 2   fecha_hora_pedido   24990 non-null  datetime64[ns]
 3   pais                24990 non-null  object        
 4   dispositivo         24990 non-null  object        
 5   fuente_referencia   24990 non-null  object        
 6   nombre_producto     24990 non-null  object        
 7   categoria_producto  24960 non-null  object        
 8   cantidad            24940 non-null  float64       
 9   precio_unitario     24940 non-null  float64       
 10  monto_descuento     24940 non-null  float64       
 11  monto_total         24990 non-null  float64       
 12  tipo_transaccion    24990 non-null  object        
dtypes: datetime64[ns](1), float64(4), obje

### Tratamiento de Valores Nulos (`NaN`) en Columnas Numéricas

En el dataset `Orders` (24,990 registros), las columnas numéricas `cantidad`, `precio_unitario` y `monto_descuento` conservan **50 valores nulos (`NaN`)** (24,940 registros no nulos). Se decidió **no eliminar estas filas ni imputar valores artificiales** por las siguientes razones analíticas y de negocio:

* **Preservación de datos transaccionales válidos:** Cada una de estas 50 filas representa una transacción real que contiene información completa y valiosa en dimensiones clave como `id_pedido`, `id_usuario`, `fecha_hora_pedido`, `pais`, `dispositivo`, `fuente_referencia` y `monto_total`.
* **Evitar sesgos en la imputación:** Reemplazar estos nulos con métricas de tendencia central (como la media o la mediana) introduciría valores ficticios en el volumen de piezas o montos de descuento, distorsionando el cálculo real del comportamiento transaccional.
* **Compatibilidad con el modelado en Power BI / DAX:**
  * **Conteo de tráfico y transacciones:** Permitirá contabilizar el volumen total de órdenes y el comportamiento por canal de adquisición sin perder registros.
  * **Cálculos agregados (`SUM`, `AVERAGE`):** Las funciones del lenguaje DAX en Power BI ignoran automáticamente los valores nulos (`BLANK`) al ejecutar sumas o promedios, por lo que las métricas financieras sobre piezas o montos se mantendrán 100% precisas sin afectar los denominadores.

In [35]:
# exportar datasets
orders.to_csv('orders_clean.csv', index=False)
catalog.to_csv('catalog_clean.csv', index=False)
marketing.to_csv('marketing_clean.csv', index=False)

---

##  Analizar si el negocio es rentable


In [36]:
# Asegurar que 'nombre_producto' en orders no tenga Nulos antes del merge
orders['nombre_producto_clean'] = orders['nombre_producto'].fillna('Sin especificar')

In [37]:
# Fila comodín en el catálogo para productos no especificados
fila_sin_especificar = pd.DataFrame([{
    'nombre_producto': 'Sin especificar',
    'categoria_producto': 'Sin especificar',
    'costo_unitario': 0.0
}])

# Concatenar al catálogo existente (sin alterar las variables originales)
catalog_prep = pd.concat([catalog, fila_sin_especificar], ignore_index=True)

# Realizar el LEFT JOIN
orders_merge_catalog = orders.merge(
    catalog_prep[['nombre_producto', 'categoria_producto', 'costo_unitario']], 
    left_on='nombre_producto_clean',
    right_on='nombre_producto', 
    how='left',
    suffixes=('', '_catalog')
)
# Imputar o rellenar los valores nulos en registros de Devolución/Sin Especificar
orders_merge_catalog['nombre_producto'] = orders_merge_catalog['nombre_producto_clean'].fillna('Sin especificar')
orders_merge_catalog['categoria_producto'] = orders_merge_catalog['categoria_producto'].fillna('Sin especificar')
orders_merge_catalog['costo_unitario'] = orders_merge_catalog['costo_unitario'].fillna(0)

orders_merge_catalog['costo_transaccion'] = (
    orders_merge_catalog['costo_unitario'] * orders_merge_catalog['cantidad']
)

orders_merge_catalog.head()


,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total,tipo_transaccion,nombre_producto_clean,nombre_producto_catalog,categoria_producto_catalog,costo_unitario,costo_transaccion
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2.0,332.69,0.0,665.37,Venta,Jacket-Winter-M,Jacket-Winter-M,Moda,189.31,378.62
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,176.86,5.0,171.86,Venta,Tablet-Standard-64GB,Tablet-Standard-64GB,Electronica,25.21,25.21
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2.0,102.99,10.0,195.99,Venta,Blender-XL-Red,Blender-XL-Red,Hogar,176.64,353.28
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electronica,1.0,257.87,15.0,242.87,Venta,Tablet-Standard-64GB,Tablet-Standard-64GB,Electronica,25.21,25.21
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1.0,336.28,0.0,336.28,Venta,Blender-XL-Red,Blender-XL-Red,Hogar,176.64,176.64


In [38]:
catalog_prep.head(10)

,nombre_producto,categoria_producto,costo_unitario,proveedor
0,Laptop-Gaming-16GB,Electronica,280.68,"Fuller, Pena and Myers"
1,Phone-Pro-128GB,Electronica,10.12,King Ltd
2,Tablet-Standard-64GB,Electronica,25.21,Bowers LLC
3,Blender-XL-Red,Hogar,176.64,Long-Reid
4,Vacuum-Pro-Black,Hogar,16.60,"Rivera, Carr and Finley"
5,Sneakers-Urban-42,Moda,17.21,Greene-Smith
6,Jacket-Winter-M,Moda,189.31,Mcmillan-Rhodes
7,Sin especificar,Sin especificar,0.00,NaN


In [39]:
# Eliminación de columna repetida y renombramiento de la que se usará en el df
orders_merge_catalog = orders_merge_catalog.drop('categoria_producto_catalog', axis=1)

In [40]:
orders_merge_catalog[orders_merge_catalog['nombre_producto']== 'Sin especificar']

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total,tipo_transaccion,nombre_producto_clean,nombre_producto_catalog,costo_unitario,costo_transaccion
44,order_44,user_2899,2025-02-22,Colombia,desktop,Sin especificar,Sin especificar,Sin especificar,1.0,318.38,5.0,313.38,Venta,Sin especificar,Sin especificar,0.0,0.0
45,order_45,user_6196,2025-02-10,Mexico,mobile,Sin especificar,Sin especificar,Sin especificar,2.0,354.06,10.0,698.12,Venta,Sin especificar,Sin especificar,0.0,0.0
46,order_46,user_5815,2025-04-24,Mexico,mobile,Sin especificar,Sin especificar,Sin especificar,1.0,359.31,10.0,349.31,Venta,Sin especificar,Sin especificar,0.0,0.0
47,order_47,user_406,2025-05-08,Colombia,mobile,Sin especificar,Sin especificar,Sin especificar,2.0,476.78,0.0,953.56,Venta,Sin especificar,Sin especificar,0.0,0.0
48,order_48,user_6187,2025-01-26,Colombia,desktop,Sin especificar,Sin especificar,Sin especificar,2.0,137.96,0.0,275.93,Venta,Sin especificar,Sin especificar,0.0,0.0
49,order_49,user_7225,2025-04-04,Argentina,desktop,Sin especificar,Sin especificar,Sin especificar,2.0,433.76,0.0,867.53,Venta,Sin especificar,Sin especificar,0.0,0.0
50,order_50,user_1840,2025-02-05,Mexico,desktop,Sin especificar,Sin especificar,Sin especificar,1.0,495.38,5.0,490.38,Venta,Sin especificar,Sin especificar,0.0,0.0
51,order_51,user_5998,2025-02-21,Mexico,desktop,Sin especificar,Sin especificar,Sin especificar,2.0,382.94,0.0,765.89,Venta,Sin especificar,Sin especificar,0.0,0.0
52,order_52,user_5679,2025-04-17,Argentina,mobile,Sin especificar,Sin especificar,Sin especificar,2.0,415.01,0.0,830.01,Venta,Sin especificar,Sin especificar,0.0,0.0
53,order_53,user_5936,2025-01-02,Argentina,desktop,Sin especificar,Sin especificar,Sin especificar,2.0,236.49,0.0,472.98,Venta,Sin especificar,Sin especificar,0.0,0.0


In [41]:
orders_merge_catalog[orders_merge_catalog['tipo_transaccion']== 'Devolucion']

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total,tipo_transaccion,nombre_producto_clean,nombre_producto_catalog,costo_unitario,costo_transaccion
266,order_266,user_7011,2025-03-13,Sin especificar,desktop,paid_search,Phone-Pro-128GB,Electronica,-2.0,101.31,10.0,-192.62,Devolucion,Phone-Pro-128GB,Phone-Pro-128GB,10.12,-20.24
267,order_267,user_1087,2025-05-07,Sin especificar,desktop,social,Phone-Pro-128GB,Electronica,-1.0,43.50,5.0,-38.50,Devolucion,Phone-Pro-128GB,Phone-Pro-128GB,10.12,-10.12
268,order_268,user_84,2025-02-19,Sin especificar,desktop,organic,Phone-Pro-128GB,Electronica,-1.0,497.65,5.0,-492.65,Devolucion,Phone-Pro-128GB,Phone-Pro-128GB,10.12,-10.12
269,order_269,user_3323,2025-05-25,Sin especificar,desktop,paid_search,Phone-Pro-128GB,Electronica,-1.0,423.53,0.0,-423.53,Devolucion,Phone-Pro-128GB,Phone-Pro-128GB,10.12,-10.12


In [42]:
# Columna agregada para cálculo del costo total
orders_merge_catalog['costo_transaccion'] = orders_merge_catalog['costo_unitario']* orders_merge_catalog['cantidad']

# KPI's de rentabilidad del negocio
ingreso_total = (orders_merge_catalog[orders_merge_catalog['tipo_transaccion']== 'Venta' ]['monto_total'].sum() )
monto_devolucion = orders_merge_catalog[orders_merge_catalog['tipo_transaccion']== 'Devolucion']['monto_total'].abs().sum()
costo_total = orders_merge_catalog['costo_transaccion'].sum()
marketing_inversion = marketing['gasto'].sum()
profit = ingreso_total - (costo_total + marketing_inversion)
margen_ganancia_general = profit / ingreso_total

# Impresión con formato a 2 decimales
print(f'Ingreso Total: ${ingreso_total:,.2f}')
print(f'Monto devoluciones: ${monto_devolucion:,.2f}')
print(f'Costo Total: ${costo_total:,.2f}')
print(f'Inversión por Marketing: ${marketing_inversion:,.2f}')
print(f'Profit: ${profit:,.2f}')
print(f'Margen de ganancia (%): {margen_ganancia_general:.2%}' )

Ingreso Total: $9,643,909.56
Monto devoluciones: $1,147.30
Costo Total: $3,828,818.41
Inversión por Marketing: $2,871,843.53
Profit: $2,943,247.62
Margen de ganancia (%): 30.52%


In [43]:
# KPI's de comportamiento de ventas
ticket_promedio = orders_merge_catalog[orders_merge_catalog['tipo_transaccion']== 'Venta' ]['monto_total'].mean()
cantidad_promedio = orders_merge_catalog['cantidad'].mean()
cantidad_transacciones = orders_merge_catalog['id_pedido'].count()

print(f'Ticket Promedio: ${ticket_promedio:,.2f}')
print(f'Volumen Promedio de Venta (pzas): {cantidad_promedio:.0f}')
print(f'Total de Transacciones: {cantidad_transacciones:.0f} transacciones')

Ticket Promedio: $385.97
Volumen Promedio de Venta (pzas): 2
Total de Transacciones: 24990 transacciones


In [44]:
# Producto más vendido (cantidad/volumen), costo e ingresos (monto_total)
resumen_productos = orders_merge_catalog.groupby('nombre_producto_clean')[['cantidad','costo_transaccion', 'monto_total']].sum(min_count= 1).sort_values(by= 'cantidad', ascending= False)

# Cálculo de ganancias por producto (sin considerar gasto por marketing, pero si devoluciones)
resumen_productos['ganancia'] = ( resumen_productos['monto_total'] ) - ( (resumen_productos['costo_transaccion']) )
resumen_productos['margen_utilidad'] = ( resumen_productos['ganancia'] ) / ( (resumen_productos['monto_total']) )

resumen_productos.loc[resumen_productos['costo_transaccion'] == 0, ['costo_transaccion', 'margen_utilidad']] = np.nan

resumen_productos.style.format({
    'cantidad': '{:,.0f} pzas',
    'costo_transaccion': '${:,.2f}',
    'monto_total': '${:,.2f}',
    'ganancia': '${:,.2f}',
    'margen_utilidad': '{:.2%}',  # Convierte 0.1542 en 15.42%
}, na_rep= "N/D") # Muestra 'N/D' (No Disponible) en lugar de 100% o $0.00 (para categorias de producto sin especificar)

,cantidad,costo_transaccion,monto_total,ganancia,margen_utilidad
nombre_producto_clean,,,,,
Vacuum-Pro-Black,"6,284 pzas","$104,314.40","$1,623,994.00","$1,519,679.60",93.58%
Blender-XL-Red,"6,279 pzas","$1,109,122.56","$1,613,576.87","$504,454.31",31.26%
Jacket-Winter-M,"6,256 pzas","$1,184,323.36","$1,610,784.01","$426,460.65",26.48%
Sneakers-Urban-42,"6,172 pzas","$106,220.12","$1,574,713.00","$1,468,492.88",93.25%
Laptop-Gaming-16GB,"4,198 pzas","$1,178,294.64","$1,086,570.21","$-91,724.43",-8.44%
Tablet-Standard-64GB,"4,153 pzas","$104,697.13","$1,063,771.00","$959,073.87",90.16%
Phone-Pro-128GB,"4,135 pzas","$41,846.20","$1,057,090.55","$1,015,244.35",96.04%
Sin especificar,45 pzas,N/D,"$12,262.62","$12,262.62",N/D


In [45]:
# Gasto de marketing por canal
resumen_marketing= marketing.groupby('canal')['gasto'].sum().sort_values(ascending= False)
# Formato
resumen_marketing = resumen_marketing.map('${:,.2f}'.format)

resumen_marketing


canal
social             $918,043.21
organic            $913,533.01
paid_search        $863,088.21
Sin especificar    $177,179.10
Name: gasto, dtype: object

# Funnel de conversión (SQL)

In [46]:
import pandas as pd
from sqlalchemy import create_engine

# ======================
# Conexión (NO modificar)
# ======================
db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-production-db-en'
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [47]:
# Explorar tabla events
# =========================
query_events = '''
SELECT *
FROM events;
'''
events = pd.read_sql(query_events, con=engine)
events.head(20)

,id_usuario,id_sesion,nombre_evento,timestamp_evento,pais,dispositivo,fuente_referencia,categoria_producto
0,user_6772,6a97f2af-32ae-4186-8c92-04025be1a27b,first_visit,2025-05-17,Colombia,desktop,organic,Moda
1,user_5883,369b767c-1c33-4b2f-a652-c7c0ef92cfc9,add_to_cart,2025-02-23,Mexico,mobile,social,Hogar
2,user_5946,60039041-e78b-474c-87b3-c0b7e9c30708,add_payment_info,2025-05-15,Colombia,desktop,social,Electronica
3,user_827,18252a64-f389-4ef7-9e58-dadad4a3491e,purchase,2025-03-31,Mexico,mobile,social,Moda
4,user_2361,221b364e-cdc5-4668-b698-18d5ba849a67,first_visit,2025-01-22,Argentina,desktop,paid_search,Electronica
5,user_4402,0892643c-f80d-4eb4-9ca8-3638b16c6ed4,begin_checkout,2025-04-21,Argentina,desktop,paid_search,Hogar
6,user_767,0005a13c-3c1f-496b-a26a-f3c282f19e13,begin_checkout,2025-03-01,Mexico,mobile,social,Electronica
7,user_2694,b07336e6-1207-4ee1-82e3-1fa7fbd194e3,add_payment_info,2025-03-22,Colombia,desktop,organic,Hogar
8,user_7209,f42bd8cd-aff2-4b7e-94e2-d4d885d23afc,add_to_cart,2025-05-07,Mexico,desktop,organic,Moda
9,user_1788,abf1af77-e96b-4490-9efb-3fa5d8271ad4,add_to_cart,2025-05-13,Argentina,desktop,organic,Moda


In [48]:
events['nombre_evento'].unique()

array(['first_visit', 'add_to_cart', 'add_payment_info', 'purchase',
       'begin_checkout', 'select_item'], dtype=object)

In [49]:
# PARTE 1: Totales del funnel
# ======================
query_totals = """
SELECT
    nombre_evento,
    COUNT(DISTINCT id_usuario) AS total_eventos
FROM events
GROUP BY nombre_evento
ORDER BY 
    CASE nombre_evento
        WHEN 'first_visit' THEN 1
        WHEN 'select_item' THEN 2
        WHEN 'add_to_cart' THEN 3
        WHEN 'begin_checkout' THEN 4
        WHEN 'add_payment_info' THEN 5
        WHEN 'purchase' THEN 6
        ELSE 7
    END;
"""
totals = pd.read_sql(query_totals, con=engine)
totals

,nombre_evento,total_eventos
0,first_visit,7796
1,select_item,7582
2,add_to_cart,7634
3,begin_checkout,7208
4,add_payment_info,6250
5,purchase,6240


### 📌 Nota sobre la Secuencia del Embudo y Anomalía de Volumen

Al analizar la secuencia natural de navegación (**`first_visit` $\rightarrow$ `select_item` $\rightarrow$ `add_to_cart` $\rightarrow$ `begin_checkout` $\rightarrow$ `add_payment_info` $\rightarrow$ `purchase`**) mediante el conteo de **usuarios únicos** (`COUNT(DISTINCT id_usuario)`), se identifica el siguiente comportamiento:

 * **`select_item`**: 7,582 usuarios únicos
* **`add_to_cart`**: 7,634 usuarios únicos

**Hallazgo:**
 Existe un superávit de **52 usuarios únicos** (+0.69%) en la etapa de `add_to_cart` en comparación con `select_item`. 

 **Explicación técnica/negocio:**
 Este comportamiento no representa un error en el procesamiento de datos, sino una característica del flujo de la plataforma: existen usuarios que ejecutan la adición al carrito de forma directa (por ejemplo, a través de botones de compra rápida en listas generales, carruseles de recomendación o enlaces externos) sin necesidad de registrar la interacción previa de selección detallada (`select_item`).
 
 *Para preservar la transparencia del reporte comercial, los datos se presentan sin imputación ni exclusión arbitraria.*

# Análisis del Embudo de Conversión (Funnel Analysis) - Rappi Pro / Rappi Plus

## 1. Resumen de Usuarios Únicos por Etapa del Embudo

| Etapa del Embudo | Nombre de Métrica (`total`) | Usuarios Únicos | Conversión Absoluta (vs. First Visit) |
| :--- | :--- | :---: | :---: |
| **1. Primera Visita** | `first_visit_total` | **7,796** | 100.00% |
| **2. Selección de Producto** | `select_item_total` | **7,582** | 97.26% |
| **3. Agregar al Carrito** | `add_to_cart_total` | **7,634** | 97.92% |
| **4. Iniciar Checkout** | `begin_checkout_total` | **7,208** | 92.46% |
| **5. Agregar Información de Pago** | `add_payment_info_total` | **6,250** | 80.17% |
| **6. Compra Finalizada** | `purchase_total` | **6,240** | 80.04% |

---

## 2. Análisis de Pérdida de Usuarios (Dropoff / Caída por Etapa)

| Transición de Etapa | Métrica de Pérdida (`dropoff_pct`) | Caída (%) | Diagnóstico y Comportamiento |
| :--- | :--- | :---: | :--- |
| **Primera Visita → Selección de Producto** | `dropoff_first_to_select_pct` | **2.74%** | Baja fricción en navegación inicial. |
| **Selección de Producto → Agregar al Carrito** | `dropoff_select_to_cart_pct` | **-0.69%** | Variación nula / compras directas desde listado. |
| **Agregar al Carrito → Iniciar Checkout** | `dropoff_cart_to_checkout_pct` | **5.58%** | Fricción ligera previa al checkout. |
| **Iniciar Checkout → Agregar Información de Pago** | `dropoff_checkout_to_payment_pct` | **13.29%** | **PUNTO CRÍTICO:** Caída principal de 1,042 usuarios al registrar/seleccionar el método de pago. |
| **Agregar Información de Pago → Compra** | `dropoff_payment_to_purchase_pct` | **0.16%** | Conversión casi perfecta (99.84%). |

---

## 3. Puntos Clave para la Entrega (Key Takeaways)

1. **Punto Principal de Pérdida de Clientes:** 
   El paso entre `begin_checkout` y `add_payment_info` representa el cuello de botella del embudo con una caída del **13.29%** (**1,042 usuarios perdidos**). Esto se debe a fricciones en la pasarela de pago (ausencia de pagos a un solo clic, costos/tarifas mostrados hasta el final o reingreso de datos de tarjeta).
2. **Eficiencia en la Conversión Final:**
   Una vez que el usuario ingresa su información de pago, el abandono es prácticamente nulo (**0.16%** / 10 usuarios), confirmando que la transacción final no tiene fallas técnicas.

In [50]:
# PARTE 2: Conversiones
# ======================

query_conversion = """
WITH cte_first_visit AS (
    SELECT COUNT(DISTINCT(id_usuario)) AS total 
    FROM events 
    WHERE nombre_evento = 'first_visit'
),
cte_select_item AS (
    SELECT COUNT(DISTINCT(id_usuario)) AS total 
    FROM events 
    WHERE nombre_evento = 'select_item'
),
cte_add_to_cart AS (
    SELECT COUNT(DISTINCT(id_usuario)) AS total 
    FROM events 
    WHERE nombre_evento = 'add_to_cart'
),
cte_begin_checkout AS (
    SELECT COUNT(DISTINCT(id_usuario)) AS total 
    FROM events 
    WHERE nombre_evento = 'begin_checkout'
),
cte_add_payment_info AS (
    SELECT COUNT(DISTINCT(id_usuario)) AS total 
    FROM events 
    WHERE nombre_evento = 'add_payment_info'
),
cte_purchase AS (
    SELECT COUNT(DISTINCT(id_usuario)) AS total 
    FROM events 
    WHERE nombre_evento = 'purchase'
)
SELECT
    -- 1. Totales de cada evento
    (SELECT total FROM cte_first_visit) AS first_visit_total,
    (SELECT total FROM cte_select_item) AS select_item_total,
    (SELECT total FROM cte_add_to_cart) AS add_to_cart_total,
    (SELECT total FROM cte_begin_checkout) AS begin_checkout_total,
    (SELECT total FROM cte_add_payment_info) AS add_payment_info_total,
    (SELECT total FROM cte_purchase) AS purchase_total,

    -- 2. Cálculo de Drop-off % (Pérdida paso a paso)
    ROUND(
        ((SELECT total FROM cte_first_visit) - (SELECT total FROM cte_select_item)) * 100.0 / 
        NULLIF((SELECT total FROM cte_first_visit), 0), 2
    ) AS dropoff_first_to_select_pct,

    ROUND(
        ((SELECT total FROM cte_select_item) - (SELECT total FROM cte_add_to_cart)) * 100.0 / 
        NULLIF((SELECT total FROM cte_select_item), 0), 2
    ) AS dropoff_select_to_cart_pct,

    ROUND(
        ((SELECT total FROM cte_add_to_cart) - (SELECT total FROM cte_begin_checkout)) * 100.0 / 
        NULLIF((SELECT total FROM cte_add_to_cart), 0), 2
    ) AS dropoff_cart_to_checkout_pct,

    ROUND(
        ((SELECT total FROM cte_begin_checkout) - (SELECT total FROM cte_add_payment_info)) * 100.0 / 
        NULLIF((SELECT total FROM cte_begin_checkout), 0), 2
    ) AS dropoff_checkout_to_payment_pct,

    ROUND(
        ((SELECT total FROM cte_add_payment_info) - (SELECT total FROM cte_purchase)) * 100.0 / 
        NULLIF((SELECT total FROM cte_add_payment_info), 0), 2
    ) AS dropoff_payment_to_purchase_pct
;
"""


from sqlalchemy import text

conversion = pd.read_sql(text(query_conversion), con=engine)
conversion

,first_visit_total,select_item_total,add_to_cart_total,begin_checkout_total,add_payment_info_total,purchase_total,dropoff_first_to_select_pct,dropoff_select_to_cart_pct,dropoff_cart_to_checkout_pct,dropoff_checkout_to_payment_pct,dropoff_payment_to_purchase_pct
0,7796,7582,7634,7208,6250,6240,2.74,-0.69,5.58,13.29,0.16


**Validación de datos del funnel: Usuarios que hicieron 'add_to_cart' pero NUNCA registraron 'select_item'**

In [51]:
query_usuarios_salto = """
WITH usuarios_cart AS (
    SELECT DISTINCT(id_usuario)
    FROM events
    WHERE nombre_evento = 'add_to_cart'
),
usuarios_select AS(
    SELECT DISTINCT(id_usuario)
    FROM events
    WHERE nombre_evento = 'select_item'
)
SELECT
    COUNT(c.id_usuario) AS usuarios_directos_carrito
FROM usuarios_cart c
LEFT JOIN usuarios_select s ON c.id_usuario = s.id_usuario
WHERE s.id_usuario IS NULL;
"""
usuarios_salto = pd.read_sql(text(query_usuarios_salto), con=engine)
usuarios_salto

,usuarios_directos_carrito
0,401


**Validación de datos del funnel: Usuarios que hicieron 'select_item' pero NUNCA registraron 'add_to_cart'**

In [52]:
# Usuarios que hicieron 'select_item' pero NUNCA llegaron a 'add_to_cart'
query_abandonos_select = """
WITH usuarios_select AS (
    SELECT DISTINCT id_usuario
    FROM events
    WHERE nombre_evento = 'select_item'
),
usuarios_cart AS (
    SELECT DISTINCT id_usuario
    FROM events
    WHERE nombre_evento = 'add_to_cart'
)
SELECT 
    COUNT(s.id_usuario) AS usuarios_que_abandonaron_en_select
FROM usuarios_select s
LEFT JOIN usuarios_cart c ON s.id_usuario = c.id_usuario
WHERE c.id_usuario IS NULL;
"""

abandonos = pd.read_sql(text(query_abandonos_select), con=engine)
abandonos

,usuarios_que_abandonaron_en_select
0,349


**Nota de comportamiento:** La diferencia neta de +52 usuarios en `add_to_cart` resulta del balance entre 401 usuarios que agregaron productos directamente al carrito sin pasar por `select_item` y 349 usuarios que abandonaron el flujo tras seleccionar un producto.

# Retención por cohortes

In [53]:
# Explorar tabla users
# =========================
query_users = '''
SELECT *
FROM users;
'''
users = pd.read_sql(query_users, con=engine)
users.head(15)

,id_usuario,fecha_registro,país,dispositivo,tipo_plan
0,user_0,2025-01-29,Mexico,mobile,free
1,user_1,2025-01-07,Mexico,mobile,free
2,user_2,2025-03-12,Argentina,mobile,free
3,user_3,2025-03-04,Mexico,desktop,free
4,user_4,2025-02-27,Argentina,desktop,free
5,user_5,2025-02-05,Colombia,desktop,free
6,user_6,2025-01-27,Mexico,mobile,free
7,user_7,2025-05-20,Colombia,mobile,free
8,user_8,2025-01-23,Mexico,mobile,free
9,user_9,2025-04-19,Colombia,desktop,free


In [54]:
# Convertir a datetime (por si viene como string) y extraer Año-Mes
pd.to_datetime(users['fecha_registro']).dt.strftime('%Y-%m').unique()

array(['2025-01', '2025-03', '2025-02', '2025-05', '2025-04'],
      dtype=object)

In [55]:
# Explorar tabla user_activity
# =========================
query_user_activity = '''
SELECT *
FROM user_activity;
'''
user_activity = pd.read_sql(query_user_activity, con=engine)
user_activity.head(15)

,id_usuario,fecha_actividad,dias_despues_registro,activo
0,user_0,2025-02-05,7,0
1,user_0,2025-02-12,14,1
2,user_0,2025-02-19,21,1
3,user_0,2025-02-26,28,0
4,user_1,2025-01-14,7,0
5,user_1,2025-01-21,14,0
6,user_1,2025-01-28,21,1
7,user_1,2025-02-04,28,0
8,user_2,2025-03-19,7,0
9,user_2,2025-03-26,14,1


In [56]:
# Convertir a datetime (por si viene como string) y extraer Año-Mes
pd.to_datetime(user_activity['fecha_actividad']).dt.strftime('%Y-%m').unique()

array(['2025-02', '2025-01', '2025-03', '2025-04', '2025-05', '2025-06'],
      dtype=object)

In [57]:
# Retención por cohortes
# ======================
# NOTA: En la línea 10 se utiliza SUBSTRING(u.fecha_registro FROM 1 FOR 7) debido a que el formato de origen 
# en la base de datos para la columna fecha_registro es texto (VARCHAR), por lo que las transformaciones 
# a datetime en Pandas solo aplican en la memoria local del notebook.

query_cohort_retention_final = '''
WITH base_retencion AS (
    SELECT 
        SUBSTRING(u.fecha_registro FROM 1 FOR 7) AS cohorte_mes, 
        u.id_usuario,
        a.dias_despues_registro,
        a.activo
    FROM users u
    LEFT JOIN user_activity a ON u.id_usuario = a.id_usuario
)
SELECT 
    cohorte_mes,
    -- 1. Tamaño base de la cohorte (Semana 0)
    COUNT(DISTINCT id_usuario) AS tamano_cohorte,

    -- 2. Conteo de usuarios retenidos activos por semana
    COUNT(DISTINCT CASE WHEN dias_despues_registro = 7 AND activo = 1 THEN id_usuario END) AS retenido_w1,
    COUNT(DISTINCT CASE WHEN dias_despues_registro = 14 AND activo = 1 THEN id_usuario END) AS retenido_w2,
    COUNT(DISTINCT CASE WHEN dias_despues_registro = 21 AND activo = 1 THEN id_usuario END) AS retenido_w3,

    -- 3. Tasa de retención semanal (%)
    ROUND(
        COUNT(DISTINCT CASE WHEN dias_despues_registro = 7 AND activo = 1 THEN id_usuario END) * 100.0 / 
        NULLIF(COUNT(DISTINCT id_usuario), 0), 2
    ) AS semana_1,

    ROUND(
        COUNT(DISTINCT CASE WHEN dias_despues_registro = 14 AND activo = 1 THEN id_usuario END) * 100.0 / 
        NULLIF(COUNT(DISTINCT id_usuario), 0), 2
    ) AS semana_2,

    ROUND(
        COUNT(DISTINCT CASE WHEN dias_despues_registro = 21 AND activo = 1 THEN id_usuario END) * 100.0 / 
        NULLIF(COUNT(DISTINCT id_usuario), 0), 2
    ) AS semana_3

FROM base_retencion
GROUP BY cohorte_mes
ORDER BY cohorte_mes;
'''

# EJECUCIÓN CORREGIDA: Usando text() y la conexión explícita (conn)
with engine.connect() as conn:
    cohorte_final = pd.read_sql(text(query_cohort_retention_final), con=conn)

cohorte_final

,cohorte_mes,tamano_cohorte,retenido_w1,retenido_w2,retenido_w3,semana_1,semana_2,semana_3
0,2025-01,1627,697,668,656,42.84,41.06,40.32
1,2025-02,1444,611,609,635,42.31,42.17,43.98
2,2025-03,1636,677,705,690,41.38,43.09,42.18
3,2025-04,1606,680,697,663,42.34,43.40,41.28
4,2025-05,1687,695,676,706,41.20,40.07,41.85


📊 **Observación Analítica: Tasa de Retención Periódica**

 En cohortes como **2025-03** y **2025-04**, la tasa de retención presenta repuntes en la **Semana 2** (subiendo de 41.38% a 43.09% y de 42.34% a 43.40%, respectivamente).
 
 **Interpretación:** 
 La métrica empleada evalúa la retención periódica (actividad en ventanas discretas de 7, 14 y 21 días). Esto evidencia un fenómeno de **reactivación de usuarios**: clientes que no registraron actividad en la primera semana posterior a su registro, pero que volvieron a interactuar en las semanas 2 o 3 debido a ciclos habituales de recompra o estímulos comerciales.

# Test estadístico: ¿La modificación en la UI del checkout impacta la **tasa de conversión de compra**?.



# Hipótesis principal (Conversión Global)
**H₀ (Hipótesis nula):** No existe una diferencia significativa en la tasa de conversión entre la versión de control y la versión de tratamiento ($p_{\text{control}} = p_{\text{tratamiento}}$).

**H₁ (Hipótesis alternativa):** Existe una diferencia estadísticamente significativa en la tasa de conversión entre la versión de control y la versión de tratamiento ($p_{\text{control}} \neq p_{\text{tratamiento}}$).
    
# Hipótesis secundaria (duración de la sesión- paramétrica)
**H₀ (Hipótesis nula):** No existe una diferencia significativa en la tasa de conversión entre la versión de control y la versión de tratamiento ($p_{\text{control}} = p_{\text{tratamiento}}$).

**H₁ (Hipótesis alternativa):** Existe una diferencia estadísticamente significativa en la tasa de conversión entre la versión de control y la versión de tratamiento ($p_{\text{control}} \neq p_{\text{tratamiento}}$).
   
**Test estadístico:** hipotesis principal- Z-Test de Proporciones  | Hipotesis secundaria- t-Student

**Nivel de significancia alpha:** 0.05

In [58]:
# Importar csv
experiment = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/experiment_checkout_ui.csv')

experiment.head(15)

,id_usuario,variante,convirtio,dispositivo,pais,duracion_sesion,timestamp
0,exp_user_0,tratamiento,0,mobile,Argentina,114.41,2025-03-28
1,exp_user_1,tratamiento,0,desktop,Mexico,170.03,2025-01-15
2,exp_user_2,control,1,mobile,Colombia,140.21,2025-03-18
3,exp_user_3,tratamiento,0,mobile,Colombia,151.45,2025-06-03
4,exp_user_4,tratamiento,0,desktop,Mexico,299.96,2025-01-12
5,exp_user_5,control,0,mobile,Mexico,206.70,2025-01-26
6,exp_user_6,control,0,desktop,Mexico,280.11,2025-04-15
7,exp_user_7,tratamiento,0,mobile,Argentina,28.47,2025-03-18
8,exp_user_8,tratamiento,0,desktop,Mexico,287.04,2025-02-05
9,exp_user_9,control,0,mobile,Argentina,258.91,2025-02-11


In [59]:
# Información del data frame
experiment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id_usuario       10000 non-null  object 
 1   variante         10000 non-null  object 
 2   convirtio        10000 non-null  int64  
 3   dispositivo      10000 non-null  object 
 4   pais             10000 non-null  object 
 5   duracion_sesion  10000 non-null  float64
 6   timestamp        10000 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 547.0+ KB


In [60]:
# Revisión de valores categoricos y de fecha
print('Total de ususarios únicos en experiment data frame:', experiment['id_usuario'].nunique() )
print('Categorías de la columna dispositivo:', experiment['dispositivo'].unique() )
print('Categorías de la columna país:', experiment['pais'].unique() )
print('Año y mes de la columna timestamp', pd.to_datetime(experiment['timestamp']).dt.strftime('%Y-%m').unique() )
print('Categorías de la columna variante:', experiment['variante'].unique() )
print('Categorías de la columna convirtio:', experiment['convirtio'].unique() )

Total de ususarios únicos en experiment data frame: 10000
Categorías de la columna dispositivo: ['mobile' 'desktop']
Categorías de la columna país: ['Argentina' 'Mexico' 'Colombia']
Año y mes de la columna timestamp ['2025-03' '2025-01' '2025-06' '2025-04' '2025-02' '2025-05']
Categorías de la columna variante: ['tratamiento' 'control']
Categorías de la columna convirtio: [0 1]


In [61]:
# Revisión rápida de outliers o valores atipicos númericos mediante .describe()
experiment['duracion_sesion'].describe()

count    10000.000000
mean       159.862439
std         81.074410
min         20.010000
25%         89.470000
50%        159.725000
75%        229.745000
max        300.000000
Name: duracion_sesion, dtype: float64

In [62]:
# Conversión absoluta y tasa de conversión por variante (control vs tratamiento)
conversion_variante = experiment.groupby('variante').agg(
    total_usuarios=('id_usuario', 'count'),
    usuarios_convirtieron=('convirtio', 'sum'),
    tasa_conversion=('convirtio', 'mean'),
    duracion_promedio=('duracion_sesion', 'mean')
).reset_index()

conversion_variante['tasa_conversion'] = (conversion_variante['tasa_conversion'] * 100).round(2)
conversion_variante['duracion_promedio'] = conversion_variante['duracion_promedio'].round(2)

print("--- RESUMEN POR VARIANTE ---")
conversion_variante


--- RESUMEN POR VARIANTE ---


,variante,total_usuarios,usuarios_convirtieron,tasa_conversion,duracion_promedio
0,control,4965,779,15.69,161.04
1,tratamiento,5035,820,16.29,158.70


In [63]:
# Tabla de contingencia: Conversión por Variante y Dispositivo
contingencia_dispositivo = experiment.groupby(['dispositivo', 'variante']).agg(
    total_usuarios=('id_usuario', 'count'),
    usuarios_convirtieron=('convirtio', 'sum'),
    tasa_conversion=('convirtio', lambda x: (x.mean() * 100).round(2)),
    duracion_promedio=('duracion_sesion', lambda x: x.mean().round(2))
).reset_index()

contingencia_dispositivo


,dispositivo,variante,total_usuarios,usuarios_convirtieron,tasa_conversion,duracion_promedio
0,desktop,control,2508,461,18.38,162.45
1,desktop,tratamiento,2534,461,18.19,157.53
2,mobile,control,2457,318,12.94,159.61
3,mobile,tratamiento,2501,359,14.35,159.88


In [64]:
# Tabla de contingencia: Conversión por Variante y País
contingencia_pais = experiment.groupby(['pais', 'variante']).agg(
    total_usuarios=('id_usuario', 'count'),
    usuarios_convirtieron=('convirtio', 'sum'),
    tasa_conversion=('convirtio', lambda x: (x.mean() * 100).round(2)),
    duracion_promedio=('duracion_sesion', lambda x: x.mean().round(2))
).reset_index()

contingencia_pais

,pais,variante,total_usuarios,usuarios_convirtieron,tasa_conversion,duracion_promedio
0,Argentina,control,1656,268,16.18,162.40
1,Argentina,tratamiento,1661,273,16.44,158.37
2,Colombia,control,1632,259,15.87,161.10
3,Colombia,tratamiento,1646,255,15.49,158.71
4,Mexico,control,1677,252,15.03,159.64
5,Mexico,tratamiento,1728,292,16.90,159.00


In [65]:
from statsmodels.stats.proportion import proportions_ztest
import scipy.stats as stats

# Extraer conteos de conversiones y totales
exitos = experiment.groupby('variante')['convirtio'].sum().loc[['tratamiento', 'control']].values
muestras = experiment.groupby('variante')['convirtio'].count().loc[['tratamiento', 'control']].values

# Ejecutar Z-test para proporciones
z_stat, p_value_conv = proportions_ztest(count=exitos, nobs=muestras, alternative='two-sided')

# Ejecutar t-Student para Duración de Sesión (Prueba Paramétrica)
duracion_control = experiment[experiment['variante'] == 'control']['duracion_sesion']
duracion_trat = experiment[experiment['variante'] == 'tratamiento']['duracion_sesion']
t_stat, p_value_dur = stats.ttest_ind(duracion_control, duracion_trat, equal_var=False)

# Imprimir Resultados
alpha = 0.05
print("=== RESULTADOS DE LA PRUEBA DE HIPÓTESIS ===")
print(f"Z-statistic (Conversión): {z_stat:.4f} | p-value: {p_value_conv:.4f}")
if p_value_conv < alpha:
    print("-> Rechazamos H0: La diferencia en conversión ES estadísticamente significativa.")
else:
    print("-> No se rechaza H0: No hay evidencia suficiente de cambio en la conversión.\n")

print(f"T-statistic (Duración): {t_stat:.4f} | p-value: {p_value_dur:.4f}")
if p_value_dur < alpha:
    print("-> Rechazamos H0: La diferencia en duración de sesión ES estadísticamente significativa.")
else:
    print("-> No se rechaza H0: La duración promedio entre grupos es estadísticamente equivalente.")

=== RESULTADOS DE LA PRUEBA DE HIPÓTESIS ===
Z-statistic (Conversión): 0.8133 | p-value: 0.4161
-> No se rechaza H0: No hay evidencia suficiente de cambio en la conversión.

T-statistic (Duración): 1.4454 | p-value: 0.1484
-> No se rechaza H0: La duración promedio entre grupos es estadísticamente equivalente.


### 📊 Resultados de las Pruebas de Hipótesis (A/B Test)

Nivel de significancia establecido: $\alpha = 0.05$

* **1. Tasa de Conversión (Z-test de 2 proporciones)**

    **Hipótesis Nula (H0):** La tasa de conversión del grupo de tratamiento es igual a la del grupo de control.  
    **Hipótesis Alternativa (H1):** La tasa de conversión del grupo de tratamiento es diferente a la del grupo de control.

    **Estadístico Z: 0.8133 | p-value: 0.4161**

    **Resultado:** Como **p-value > 0.05**, no se rechaza la hipótesis nula **(H0)**. No existe evidencia estadística suficiente de una diferencia significativa en la tasa de conversión entre el grupo de control **(15.69%)** y tratamiento **(16.29%)**.

* **2. Duración de la Sesión (Prueba t de Student independiente)**

    **Hipótesis Nula (H0):** La duración promedio de la sesión en el grupo de tratamiento es igual a la del grupo de control.  
    **Hipótesis Alternativa (H1):** La duración promedio de la sesión en el grupo de tratamiento es diferente a la del grupo de control.

    **Estadístico t: 1.4454 | p-value: 0.1484**

    **Resultado:** Como **p-value > 0.05**, no se rechaza la hipótesis nula **(H0)**. La duración promedio de la sesión (**161.04 s** en control vs **158.70 s** en tratamiento) es estadísticamente equivalente entre ambos grupos.

**Conclusión General del Experimento:**  
Ninguna de las dos métricas evaluadas presenta cambios estadísticamente significativos. Las diferencias observadas en los datos entran dentro del margen de variabilidad aleatoria propio del muestreo.